Binding fix

efischer19 opened the pull request:
## [TNL-4152](https://openedx.atlassian.net/browse/TNL-4152)

~~A bug was discovered in the last RC, namely, we weren't binding xblocks~~
Update: everything that was here before is outdated. [ARCH-18](https://openedx.atlassian.net/browse/ARCH-18) for background.

We need to:
- pass `requesting_user` to any xblocks that define the `clear_student_state` method defined in `reset_student_attempts`
- change the signature of `reset_student_attempts` to accept this information
- make sure that this information is provided anytime this method is called

Sibling changes:
- https://github.com/edx/edx-proctoring/pull/281
- https://github.com/edx/edx-ora2/pull/891

edx-webhook commented:
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

efischer19 commented on lms/djangoapps/instructor/enrollment.py:
```diff
@@ -219,9 +220,11 @@ def reset_student_attempts(course_id, student, module_state_key, delete_module=F
+    user_id = anonymous_id_for_user(student, course_id)
+    submission_cleared = False
     try:
+        block = get_module_for_student(context_user, module_state_key)
```
This change is the key of this PR, all the blocks of code below are the ORA delete state code changes that got reverted when this bug was found.

efischer19 commented:
I have this working locally. I'm going to try and get a sandbox built to confirm behavior there before proceeding with reviews. Once I've confirmed it works there, I'll get reviews, rebase on master after this week's release merges, and begin orchestrating a 3-repo simultaneous release with the sibling PRs linked above.

efischer19 commented:
Tests are passing, and sandbox is up at https://ora2clearstate.sandbox.edx.org/. @robrap and @dianakhuang, can you review this along with https://github.com/edx/edx-ora2/pull/891?

I'm going to try and find out if/where this code may also be getting called, but I feel like we're pretty safe due to the "normal" way of accessing this functionality being through the instructor API, where we have a request object. Any other uses (such as the InstructorService used by proctoring) should be fairly exceptional and easy to spot.

douglashall commented on lms/djangoapps/instructor/services.py:
```diff
@@ -27,9 +27,9 @@ class InstructorService(object):
+    def delete_student_attempt(self, student_identifier, course_id, content_id, requesting_user):
         """
-        Deletes student state for a problem.
+        Deletes student state for a problem. requsting_user may be kept as an audit trail.
```
Typo: missing 'e' in 'requsting_user'

efischer19 commented:
Aha! I'm now fully confident that we've found all the uses of this code, as I'm able to search every `edx`-owned repository at the same time:
- https://github.com/search?utf8=%E2%9C%93&q=user%3Aedx+reset_student_attempts&type=Code&ref=searchresults
- https://github.com/search?q=user%3Aedx+delete_student_attempt&type=Code&utf8=%E2%9C%93

If there are any other outstanding usages, I would argue that they should be going through the API, which will handle this update to the `reset_student_attempts` method seamlessly.

robrap commented:
jenkins run bokchoy

robrap commented on lms/djangoapps/instructor/enrollment.py:
```diff
@@ -219,26 +219,41 @@ def reset_student_attempts(course_id, student, module_state_key, delete_module=F
+    # Reset the student's score in the submissions API, if xblock.clear_student_state has not done so already.
+    # TODO: Remove this once we've finalized and communicated how xblocks should handle clear_student_state
+    # and made sure that other xblocks relying on the submission api understand this is going away.
+    # We need to do this before retrieving the `StudentModule` model, because a score may exist with no student module.
```
Can you move the "TODO: ..." below the line starting "We need to..."?  Also, can you update the TODO as follows?:
```
# TODO: Should the LMS know about sub_api and call this reset, or should it generically call it on all of its xblock services as well?  See JIRA ARCH-26.
```

efischer19 replied:
:heavy_check_mark:

robrap commented on lms/djangoapps/instructor/services.py:
```diff
@@ -63,7 +63,13 @@ def delete_student_attempt(self, student_identifier, course_id, content_id):
+                    student,
+                    module_state_key,
+                    delete_module=True,
+                    requesting_user=requesting_user,
```
I think this was added as the second to last param, and not the last param.  Assuming this is a bug, can you determine why there wasn't a test failure?

robrap commented on lms/djangoapps/instructor/tests/test_enrollment.py:
```diff
@@ -389,7 +389,7 @@ def test_delete_student_attempts(self):
                 module_state_key=msk
             ).count(), 1)
-        reset_student_attempts(self.course_key, self.user, msk, delete_module=True)
+        reset_student_attempts(self.course_key, self.user, msk, delete_module=True, requesting_user=self.user)
```
Nit: I'd prefer params are ordered as they are declared if there is no reason not to, but it is fine as is.

robrap commented on lms/djangoapps/instructor/tests/test_enrollment.py:
```diff
@@ -425,7 +425,8 @@ def test_delete_submission_scores(self, _lti_mock):
         reset_student_attempts(
             self.course_key, user, problem_location,
-            delete_module=True
+            delete_module=True,
```
Same nit

robrap commented on lms/djangoapps/instructor/tests/test_enrollment.py:
```diff
@@ -478,7 +479,13 @@ def test_delete_submission_scores_attempts_children(self):
+            self.user,
+            self.parent.location,
+            delete_module=True,
+            requesting_user=self.user
```
Nit: order.

robrap commented on lms/djangoapps/instructor/tests/test_services.py:
```diff
@@ -67,7 +67,8 @@ def test_reset_student_attempts_delete(self):
             unicode(self.course.id),
-            self.problem_urlname
+            self.problem_urlname,
+            requesting_user=self.student,
```
Shouldn't this be a different user?

robrap commented on lms/djangoapps/instructor/tests/test_services.py:
```diff
@@ -88,7 +89,8 @@ def test_reset_bad_content_id(self):
             unicode(self.course.id),
-            'foo/bar/baz'
+            'foo/bar/baz',
+            requesting_user=self.student.username,
```
Shouldn't this be a different user?  Also, not a username?  Is there an assertion that should fail for this?

efischer19 replied:
As far as using a different user, this is the same idea as the ora2 tests - the test is checking if a user can delete themselves. So it can be a different user, it just isn't in this case.

It should definitely not be a username, the reason that wasn't failing is that these particular examples are negative tests for failures in the service before even getting to the `reset_student_attempts` call that uses this parameter. I believe they're just holdover code from the development process that I missed updating due to the not-failing. I'll clean them up.

robrap commented on lms/djangoapps/instructor/tests/test_services.py:
```diff
@@ -100,7 +102,8 @@ def test_reset_bad_user(self):
             unicode(self.course.id),
-            'foo/bar/baz'
+            'foo/bar/baz',
+            requesting_user=self.student.username,
```
Same assertion question.

robrap commented on lms/djangoapps/instructor/tests/test_services.py:
```diff
@@ -112,7 +115,8 @@ def test_reset_non_existing_attempt(self):
             unicode(self.course.id),
-            self.other_problem_urlname
+            self.other_problem_urlname,
+            requesting_user=self.student,
```
Again, should this be a different user?

robrap commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1980,7 +1980,13 @@ def reset_student_attempts(request, course_id):
+                student,
+                module_state_key,
+                requesting_user=request.user,
+                delete_module=delete_module
```
Same order nit.

efischer19 replied:
This one looks to be in the correct order

robrap commented:
@efischer19 Looks good.  Some minor test issues and some extremely minor nits.  You can decide whether you will fix the nits.

efischer19 commented:
@robrap I've addressed your feedback on this, https://github.com/edx/edx-proctoring/pull/281, and https://github.com/edx/edx-ora2/pull/891. Can you give them another look when you get a chance?

dianakhuang commented:
:+1: from me. Looks good!

robrap commented:
:+1:   Thanks.

efischer19 commented:
jenkins run all

efischer19 commented:
jenkins run bokchoy

efischer19 commented:
jenkins run python

efischer19 commented:
bok-choy tests are failing for the past two runs on proctoring changes, I'll investigate.
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/15298/
- https://build.testeng.edx.org/job/edx-platform-bok-choy-pr/15299/

robrap commented:
Thanks.  Got it.
